src/components/implementation/no_interface/llbooter/boot_deps.h

```diff
@@ -135,229 +189,391 @@ boot_newcomp_sinv_alloc(spdid_t spdid)
+	struct cos_aep_info  *child_aep = boot_spd_initaep_get(spdid);
+	int ret, i;
+	
+	/* FIXME: not everyone should have it. but for now, because getting cpu cycles uses this */
+	ret = cos_cap_cpy_at(ci, BOOT_CAPTBL_SELF_INITHW_BASE, boot_info, BOOT_CAPTBL_SELF_INITHW_BASE);
+	assert(ret == 0);
+
+	if (resmgr_spdid && capci->is_sched) {
+		/*
+		 * FIXME:
+		 * This is an ugly hack to allow components to do cos_introspect()
+		 * - to get thdid
+		 * - to get budget on tcap
+		 * - other introspect...requirements..
+		 *
+		 * I don't know a way to get away from this for now! 
+		 * If it were just thdid, resmgr could have returned the thdids!
+		 */
+		ret = cos_cap_cpy_at(ci, BOOT_CAPTBL_SELF_CT, boot_info, ci->captbl_cap);
+		assert(ret == 0);
+	}
 
-	cc = cos_comp_alloc(boot_info(), ct, pt, (vaddr_t)new_comp_cap_info[spdid].upcall_entry);
+	if (resmgr_spdid == spdid || !resmgr_spdid) {
+		/*
+		 * if this is a resmgr, let it manage resources.
+		 * if there is no resmgr in the system, allow every component to manage its resources.
+		 */
+		ret = cos_cap_cpy_at(ci, BOOT_CAPTBL_SELF_PT, boot_info, ci->pgtbl_cap);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(ci, BOOT_CAPTBL_SELF_CT, boot_info, ci->captbl_cap);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(ci, BOOT_CAPTBL_SELF_COMP, boot_info, ci->comp_cap);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(ci, BOOT_CAPTBL_SELF_UNTYPED_PT, boot_info, ci->mi.pgtbl_cap);
+		assert(ret == 0);
+	}
 
-	assert(cc);
-	/* manually laying out struct due to upcall addr calculation */
-	boot_newcomp_definfo_init(spdid, cc, comp_flags);
+	/* If booter should create the init caps in that component */
+	if (capci->parent_spdid == 0) {
+		boot_newcomp_defcinfo_init(spdid);
+
+		ret = cos_cap_cpy_at(ci, BOOT_CAPTBL_SELF_INITTHD_BASE, boot_info, child_aep->thd);
+		assert(ret == 0);
 
-	/* Create sinv capability from Components to Booter */
-	sinv = cos_sinv_alloc(boot_info(), boot_info()->comp_cap, (vaddr_t)__inv_test_entry,
-			(unsigned long)spdid);
+		if (capci->is_sched) {
+			ret = cos_cap_cpy_at(ci, BOOT_CAPTBL_SELF_INITRCV_BASE, boot_info, child_aep->rcv);
+			assert(ret == 0);
+			ret = cos_cap_cpy_at(ci, BOOT_CAPTBL_SELF_INITTCAP_BASE, boot_info, child_aep->tc);
+			assert(ret == 0);
+		}
 
-	assert(sinv);
+		boot_comp_sched_set(spdid);
+	}
+}
 
-	cos_cap_cpy_at(new_comp_cap_info[spdid].compinfo, BOOT_CAPTBL_SINV_CAP, boot_info(), sinv);
+static void
+boot_newcomp_create(spdid_t spdid, struct cos_compinfo *comp_info)
+{
+	struct cos_compinfo *compinfo  = boot_spd_compinfo_get(spdid);
+	struct cos_compinfo *boot_info = boot_spd_compinfo_get(0);
+	captblcap_t ct = compinfo->captbl_cap;
+	pgtblcap_t  pt = compinfo->pgtbl_cap;
+	compcap_t   cc;
+	sinvcap_t   sinv;
+	thdcap_t    main_thd;
+	int         i = 0;
+	unsigned long token = (unsigned long) spdid;
+	int ret;
 
-	boot_newcomp_sinv_alloc(spdid);
+	cc = cos_comp_alloc(boot_info, ct, pt, (vaddr_t)new_comp_cap_info[spdid].upcall_entry);
+	assert(cc);
```

Comment:
if this code is now shared with the resource manager (perhaps it is not), then we need to start doing real error checking...

Reply:
It is not.